src/main/java/plugily/projects/villagedefense/handlers/hologram/LeaderboardHologram.java

```diff
@@ -68,74 +77,31 @@ public void run() {
-    appendHoloText(header);
-
-    int limit = topAmount;
     java.util.Map<UUID, Integer> values = StatsStorage.getStats(statistic);
     List<UUID> reverseKeys = new ArrayList<>(values.keySet());
-
     Collections.reverse(reverseKeys);
 
-    for(UUID key : reverseKeys) {
-      if(limit == 0) {
-        break;
-      }
-
-      String format = color(plugin.getLanguageConfig().getString(LanguageMessage.HOLOGRAMS_FORMAT.getAccessor()));
-      format = StringUtils.replace(format, "%place%", Integer.toString((topAmount - limit) + 1));
-      format = StringUtils.replace(format, "%nickname%", getPlayerNameSafely(key));
-      format = StringUtils.replace(format, "%value%", String.valueOf(values.get(key)));
-      appendHoloText(format);
-      limit--;
-    }
-
-    if(limit > 0) {
-      for(int i = 0; i < limit; limit--) {
-        String format = color(plugin.getLanguageConfig().getString(LanguageMessage.HOLOGRAMS_FORMAT_EMPTY.getAccessor()));
-        format = StringUtils.replace(format, "%place%", Integer.toString((topAmount - limit) + 1));
-        appendHoloText(format);
+    for (int i = 0; i < lines.size(); i++) {
+      TextLine textLine = lines.get(i);
+      String text;
+      if (i < reverseKeys.size()) {
+        UUID uuid = reverseKeys.get(i);
+        text = color(plugin.getLanguageConfig().getString(LanguageMessage.HOLOGRAMS_FORMAT.getAccessor()));
+        text = StringUtils.replace(text, "%nickname%", getPlayerNameSafely(uuid));
+        text = StringUtils.replace(text, "%value%", String.valueOf(values.get(uuid)));
+      } else {
+        text = color(plugin.getLanguageConfig().getString(LanguageMessage.HOLOGRAMS_FORMAT_EMPTY.getAccessor()));
       }
+      text = StringUtils.replace(text, "%place%", Integer.toString(i + 1));
+      textLine.setText(text);
     }
   }
 
   @Override
-  public void cancel() {
+  public synchronized void cancel() {
     super.cancel();
-
-    submitSync(() -> {
-      hologram.delete();
-      return 1;
-    });
-  }
-
-  // We should perform hologram api in synchronous thread to do not cause "async catchop" problems
-  // See this method:
-  // github.com/filoghost/HolographicDisplays/blob/af038ac93f7a0d5c1d5a7abfa4a08176b9765d16/Plugin/src/main/java/com/gmail/filoghost/holographicdisplays/object/CraftHologram.java#L179
-  private void appendHoloText(final String text) {
-    submitSync(() -> {
-      hologram.appendTextLine(text);
-      return 1;
-    });
-  }
-
-  private Integer submitSync(Supplier<Integer> sup) {
-    if (!plugin.getServer().isPrimaryThread()) { // Check if current thread is async
-      try {
-        return plugin.getServer().getScheduler().callSyncMethod(plugin, () -> sup.get()).get();
-      } catch (InterruptedException | ExecutionException e) {
-        e.printStackTrace();
-      }
-    }
-
-    return sup.get();
+    lines.clear();
+    hologram.delete();
```

Comment:
Does the `delete` method not have to be in synchronous?

Reply:
It doesn't have to. `synchronized` will force the other "Hologram" threads to wait for the `delete` method to be done.
In theory, this is `synchronized`, so no problem here.